src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs

```diff
@@ -184,19 +185,49 @@ private async Task<IReadOnlyList<string>> GetResourceIdsForLogAnalyticsAsync(IRe
             }
             else
             {
-                // Since we do not know where the telemetry of each resource is, create a client that queries all workspaces in the subscription.
+                // The workspaces associated with each resource have to be retrieved.
                 var workspacesList = new List<ResourceIdentifier>();
-                foreach (string subscriptionId in resources.Select(resource => resource.SubscriptionId).Distinct(StringComparer.CurrentCultureIgnoreCase))
+
+                foreach (ResourceIdentifier resource in resources)
                 {
-                    // Try to get the workspaces list from the cache, and if it isn't there, use the Azure Resource Manager client
-                    IList<ResourceIdentifier> subscriptionWorkspaces;
-                    if (!this.subscriptionIdToWorkspaces.TryGetValue(subscriptionId, out subscriptionWorkspaces))
+                    // There is no general way to get only the specific workspace containing the telemetry for any resource.
+                    // A method that works for the Azure kubernetes Service is implemented here. If a resource is not AKS then all workspaces in it's subscription must be retrieved.
+                    if (resource.ResourceType == ResourceType.KubernetesService)
                     {
-                        subscriptionWorkspaces = await this.azureResourceManagerClient.GetAllResourcesInSubscriptionAsync(subscriptionId, new[] { ResourceType.LogAnalytics }, cancellationToken);
-                        this.subscriptionIdToWorkspaces[subscriptionId] = subscriptionWorkspaces;
+                        // Retrieve the specific workspace the target AKS cluster belongs to.
+                        ResourceIdentifier workspace;
+                        if (!this.aksClusterIdToWorkspaces.TryGetValue(resource.ToString(), out workspace))
+                        {
+                            // Try to get the workspaces from the cache, and if it isn't there, use the Azure Resource Manager client
+                            ResourceProperties querryResult = await this.azureResourceManagerClient.GetResourcePropertiesAsync(resource, cancellationToken);
+                            if (querryResult.Properties["addonProfiles"]["omsagent"]["enabled"].ToObject<bool>())
```

Comment:
>Properties["addonProfiles"]["omsagent"]["enabled"] [](start = 45, length = 50)

You need to handle nulls here (unless you really want to throw a NullReferenceException in this case). Consider:
Properties?["addonProfiles"]?["omsagent"]?["enabled"]?.ToObject<bool>() ?? false